src/back/glsl/mod.rs

```diff
@@ -1342,6 +1342,74 @@ impl<'a, W: Write> Writer<'a, W> {
                 }
                 write!(self.out, ")")?;
             }
+            // Query translates into one of the:
+            // - textureSize/imageSize
+            // - textureQueryLevels
+            // - textureSamples/imageSamples
+            Expression::ImageQuery { image, query } => {
+                // This will only panic if the module is invalid
+                let (dim, class) = match ctx.typifier.get(image, &self.module.types) {
+                    TypeInner::Image {
+                        dim,
+                        arrayed: _,
+                        class,
+                    } => (dim, class),
+                    _ => unreachable!(),
+                };
+                let components = match dim {
+                    crate::ImageDimension::D1 => 1,
+                    crate::ImageDimension::D2 => 2,
+                    crate::ImageDimension::D3 => 3,
+                    crate::ImageDimension::Cube => 2,
+                };
+                match query {
+                    crate::ImageQuery::Size { level } => match class {
+                        ImageClass::Sampled { .. } | ImageClass::Depth => {
+                            let selector = ["x", "xy", "xyz"];
+                            write!(self.out, "textureSize(")?;
+                            self.write_expr(image, ctx)?;
+                            write!(self.out, ",")?;
+                            if let Some(expr) = level {
+                                self.write_expr(expr, ctx)?;
+                            } else {
+                                write!(self.out, "0",)?;
+                            }
+                            write!(self.out, ").{}", selector[components])?;
```

Comment:
This needs to be changed because `D1` images return `int`, `D2` return an `ivec2` and `D3` would go out of bounds